packages/playwright-test/src/reporters/azure.ts

```diff
@@ -0,0 +1,481 @@
+}
+
+class AzureDevOpsReporter implements Reporter {
+  private testApi!: Test.ITestApi;
+  private coreApi!: ICoreApi;
+  private _options: AzureReporterOptions;
+  private publishedResultsCount = 0;
+  private resultsToBePublished: string[] = [];
+  private connection!: WebApi;
+  private orgUrl!: string;
+  private projectName!: string;
+  private environment?: string;
+  private planId = 0;
+  private runId!: number;
+  private logging = true;
+  private isDisabled = false;
+  private testRunTitle = '';
+  private uploadAttachments = false;
+  private attachmentsType?: TAttachmentType;
+  private token: string = '';
+
+  public constructor(options: AzureReporterOptions = {} as AzureReporterOptions) {
+    this._options = options;
+
+    this.orgUrl = this._options.orgUrl;
+    this.projectName = this._options.projectName;
+    this.environment = this._options.environment || undefined;
+    this.planId = this._options.planId;
+    this.logging = (typeof this._options.logging === 'undefined') ? true : this._options.logging;
+    this.isDisabled = this._options.isDisabled || false;
+    this.testRunTitle = `${this.environment ? `[${this.environment}]:` : ''} ${this._options.testRunTitle || 'Playwright Test Run'}` ||
+      `${this.environment ? `[${this.environment}]:` : ''}Test plan ${this.planId}`;
+    this.uploadAttachments = this._options.uploadAttachments || false;
+    this.attachmentsType = this._options.attachmentsType;
+    this.token = this._options.token;
+
+    if (!this.orgUrl || this.orgUrl.length === 0) {
+      this.log(colors.yellow("'orgUrl' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.projectName || this.projectName.length === 0) {
+      this.log(colors.yellow("'projectName' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.planId || this.planId === 0) {
+      this.log(colors.yellow("'planId' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+      return;
+    }
+    if (!this.token || this.token.length === 0) {
+      this.log(colors.yellow("'token' is not set. Reporting is disabled."));
+      this.isDisabled = true;
+    }
+    if (this.uploadAttachments) {
+      if (!this.attachmentsType) {
+        this.log(colors.yellow("'attachmentsType' is not set. Attachments Type will be set to 'screenshot' by default."));
+        this.attachmentsType = ['screenshot'];
+      }
+    }
+    this.connection = new azdev.WebApi(this.orgUrl, azdev.getPersonalAccessTokenHandler(this.token));
+  }
+
+  async onBegin(): Promise<void> {
+    if (this.isDisabled)
+      return;
+
+    this.testApi = await this.connection.getTestApi();
+
+    this.createRun(this.testRunTitle).then(run => {
+      if (run) {
+        this.runId = run.id;
+        this.log(colors.green(`Using run ${this.runId} to publish test results`));
+      } else {
+        this.isDisabled = true;
+      }
+    });
```

Comment:
```suggestion
    const run = await this.createRun(this.testRunTitle)
    if (run) {
      this.runId = run.id;
      this.log(colors.green(`Using run ${this.runId} to publish test results`));
    } else {
      this.isDisabled = true;
    }
```

try always avoid `.then` uses since it makes it not readable.